src/charm.py

```diff
@@ -9,19 +9,133 @@
+        self._configure()
+
+    def _on_config_changed(self, _: ConfigChangedEvent) -> None:
+        """Reconfigure everything on config changed."""
+        self._configure()
+
+    def _configure(self) -> None:
+        """Reconfigure everything.
+
+        This is an idempotent method to configure everything.
+        """
+        data = self._get_keystone_data()
+
+        if data:
+            logger.info("Keystone credentials are available, starting services.")
+            self.write_cloud_config(data)
+            # TODO: when openstack-exporter service is added, restart it here
+        else:
+            logger.info("Keystone credentials are not available, stopping services.")
+            # TODO: when openstack-exporter service is added, stop it here
+
+    def _is_keystone_data_ready(self, data: dict[str, str]) -> bool:
+        """Check if all the data is available from keystone.
+
+        Data is validated as unit relation data from the keystone-admin interface.
+        """
+        return all(
+            data.get(x)
+            for x in [
+                "service_protocol",
+                "service_hostname",
+                "service_port",
+                "service_username",
+                "service_password",
+                "service_project_name",
+                "service_project_domain_name",
+                "service_user_domain_name",
+                "service_region",
+            ]
+        )
+
+    def write_cloud_config(self, data: dict[str, str]) -> None:
+        """Build a standard clouds.yaml given the keystone credentials data.
+
+        This is used by the exporter to connect to the openstack cloud,
+        including credentials, keystone endpoint, ca certificate, region.
+
+        Only api version 3 is supported,
+        since v2 was removed a long time ago (Queens release)
+        https://docs.openstack.org/keystone/latest/contributor/http-api.html
+        """
+        auth_url = "{protocol}://{hostname}:{port}/v3".format(
+            protocol=data["service_protocol"],
+            hostname=data["service_hostname"],
+            port=data["service_port"],
+        )
+        contents = {
+            "clouds": {
+                CLOUD_NAME: {
+                    "region_name": data["service_region"],
+                    "identity_api_version": "3",
+                    "identity_interface": "internal",
+                    "auth": {
+                        "username": data["service_username"],
+                        "password": data["service_password"],
+                        "project_name": data["service_project_name"],
+                        "project_domain_name": data["service_project_domain_name"],
+                        "user_domain_name": data["service_user_domain_name"],
+                        "auth_url": auth_url,
+                    },
+                    "verify": data["service_protocol"] == "https",
+                    "cacert": self.config["ssl_ca"] or None,
```

Comment:
`self.config["ssl_ca"] or None` should be changed to `self.config.get("ssl_ca", "")` since the default value of that config is `unset` which is causes 

```
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/./src/charm.py", line 240, in <module>
    ops.main(OpenstackExporterOperatorCharm)
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/venv/ops/main.py", line 471, in __call__
    return main(charm_class, use_juju_for_storage=use_juju_for_storage)
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/venv/ops/main.py", line 454, in main
    framework.reemit()
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/venv/ops/framework.py", line 875, in reemit
    self._reemit()
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/venv/ops/framework.py", line 955, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/./src/charm.py", line 210, in _on_config_changed
    self.configure(event)
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/./src/charm.py", line 160, in configure
    self.write_cloud_config(data)
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/./src/charm.py", line 107, in write_cloud_config
    "cacert": self.config["ssl_ca"] or None,
  File "/var/lib/juju/agents/unit-openstack-exporter-3/charm/venv/ops/model.py", line 796, in __getitem__
    return self._data[key]
KeyError: 'ssl_ca'
```

Reply:
oh, why not `self.config.get("ssl_ca")` without `or None`

Reply:
I don't get it.. Isn't the one you describing the same as `self.config.get("ssl_ca")`?

If `ssl_ca` is unset, set to null, or empty string, `self.config.get("ssl_ca")` will return `None`

Reply:
that also works, thanks!